README.md

```diff
@@ -1,14 +1,333 @@
+   - `204` – No content.  All OK, but no data to return
+   - `100` – chunking – all OK so far, data returned in message, but there may be more
+ - `headers`
+   - `X-Reply-Address` – contains the address of a reply (relevant for query only).  Contents should be obtained from the `messageHeaders` of the query message, specifically the header named `REPLY_ADDR_HEADER`.
+   - `body` – the actual JSON body of the response.  For query, this can be
+     - Map – the map represents a “row” in a select statement in VAIL
+     - Array of Maps – this represents a bunch of rows.
+
+If the status returned is an error, the reply address header & body should still be returned. 
+Failure to return this will result in the query (on the VANTIQ) side waiting for the query to time out. 
+
+In the case of an error, 
+the body should contain an array of objects, each of which contains the following properties:
+
+ - `messageCode`: A error code that might be used for message lookup or categorization.  This is a string, and it is generally specific to the server or class therein.
+ - `messageTemplate`: A string that is the (in this case) the error.  Places where parameters should be substituted are represented by {#}, where the # is replaced by the number of the parameters array (see below), numbered from 0.
+ - `parameters`: An array of parameters for the messageTemplate.
+
+An example might be
+
+```
+[
+    { “messageCode”: “com.example.sources.mysource.invalidQuery”,
+      “messageTemplate”: 
+      “The query: {0} is invalid for sources of type {1}”,
+      “parameters”: [“hi there”, “mysource”]
+    }
+]
+```
+
+### Overall Protocol
+
+The following diagram represents the overall flow of messages between an extension source and VANTIQ.
+
+![Extension Source Protocol Flow Diagram](docs/images/ProtocolDiagram.png)
+
+## <a id="defining_a_type"></a> Defining a Type/Implementation
+
+### Creating the Source Type
+
+When creating an OPC UA Extension source, you must first create the OPC UA source type or implementation. This is done by defining a file containing a JSON document with the following properties.
+
+  - `name` -- the name of the source type you wish to create
+  - `baseType` -- the string `EXTENSION`
+  - `verticle` -- the string `service:extensionSource`
+  - `config` -- an empty JSON object
+
+To make the source type known to VANTIQ, use the vantiq cli command
+
+
+For example, suppose we are creating an extension source of type "EXAMPLE".
+To do so, we would first create the file `exampleImpl.json` with the following contents.
+
+```
+{
+   "name" : "EXAMPLE",
+   "baseType" : "EXTENSION",
+   "verticle" : "service:extensionSource",
+   "config" : {}
+}
+```
+
+Then, we would us the VANTIQ CLI to load that source implementation.
+
+```
+vantiq -s <profileName> load sourceimpls exampleImpl.json
+```
+
+where `<profileName>` is replaced by the VANTIQ profile name, and `<fileName>` identifies the file containing the definition to be loaded.
```

Comment:
`exampleImpl.json` is used in the example instead of `<fileName>`. Looks like copy-paste error

Reply:
Only one typo?  Not bad!